pkg/service/schema.go

```diff
@@ -26,47 +32,143 @@ type bindingSchema struct {
 }
 
 type parametersSchema struct {
-	Schema     string                      `json:"$schema"`
-	Type       string                      `json:"type"`
-	Properties map[string]*ParameterSchema `json:"properties,omitempty"`
-	Required   []string                    `json:"required,omitempty"`
+	Schema           string `json:"$schema"`
+	*parameterSchema `json:",inline"`
 }
 
-// ParameterSchema represents the individual schema for a given parameter
-type ParameterSchema struct {
-	Type        string                      `json:"type"`
-	Required    bool                        `json:"-"`
-	Description string                      `json:"description,omitempty"`
-	Default     interface{}                 `json:"default,omitempty"`
-	Items       *ParameterSchema            `json:"items,omitempty"`
-	Properties  map[string]*ParameterSchema `json:"properties,omitempty"`
-	Additional  *ParameterSchema            `json:"additionalProperties,omitempty"`
+type parameterSchema struct {
+	Type          string                      `json:"type"`
+	RequiredField bool                        `json:"-"`
+	Required      []string                    `json:"required,omitempty"`
+	Description   string                      `json:"description,omitempty"`
+	Default       interface{}                 `json:"default,omitempty"`
+	Items         *parameterSchema            `json:"items,omitempty"`
+	Properties    map[string]*parameterSchema `json:"properties,omitempty"`
+	Additional    *parameterSchema            `json:"additionalProperties,omitempty"` // nolint: lll
+	AllowedValues interface{}                 `json:"enum,omitempty"`
 }
 
-func (s *parametersSchema) addParameters(
-	newParams map[string]*ParameterSchema,
-) {
-	if s.Properties == nil {
-		s.Properties = newParams
-		for key, param := range newParams {
-			if param.Required {
-				s.Required = append(s.Required, key)
-			}
+// ParameterSchema defines an interface representing a given Parameter schema.
+// This could be a provision or binding parameter.
+type ParameterSchema interface {
```

Comment:
Can you explain the shift to this being an interface. I'm sure there is a good reason, but it's lost on me at the moment.

Reply:
There were two reasons:

1.) I thought we wanted to not export any of it in the first pass at adding the schema stuff
2.) I found it useful to force people to create these through the constructor like methods (i.e. an object has properties, but an array has a field that defines what the items in the array are) so the structure is a little complicated. I thought that the constructor like methods hid some of that complexity and took some of the conceptual burden off the module writer. 

Can go back to exposing the struct if you'd prefer though, I think the struct had benefits as well (like reading the code was a little clearer because of the struct literals)